deliver/lib/deliver/upload_screenshots.rb

```diff
@@ -27,8 +27,11 @@ def upload(options, screenshots, max_n_threads = 16)
 
       localizations = version.get_app_store_version_localizations
 
+      updated_sets_per_language = get_updated_screenshots(localizations, screenshots_per_language)
+      checksums_to_delete = get_checksums_to_delete(localizations, updated_sets_per_language)
+
       if options[:overwrite_screenshots]
-        delete_screenshots(localizations, screenshots_per_language, max_n_threads)
+        delete_screenshots(localizations, checksums_to_delete, max_n_threads)
```

Comment:
Hmm... this doesn't sound right to me 🤔 when a screenshot has been updated, we should always delete it, regardless of whether the `:overwrite_screenshots` field has been set. Otherwise, we'll end up with zombie screenshots! 🧟 

What I'd suggest to do instead is this:

- if `:overwrite_screenshots` is set, delete everything (as we did before) => this is essentially the "nuke" option
- otherwise, detect changes and delete only the screenshots that have changed

Reply:
I disagree with the proposed logic since then `:overwrite_screenshots` would just affect how you do the operation (whether you will delete everything and upload everything or just replace the modified and add new screenshots) while the end result stays the same. Originally, I believe `:overwrite_screenshots` was meant to mean whether you want to delete anything at all or do you just want to add additional screenshots to the already existing ones.

If we want to change the meaning of the flag in order to suit our logic here, I am all up for it. However, then, we are straying even further from ~God~ the original Fastlane implementation and submitting it to be merged (if we want to do that) would be more difficult. If we want to stay true to the original, I'd argue that if `:overwrite_screenshots` is set to `false` then we detect all updated/new screenshots and attempt to append them at the back of the screenshot sets.

Reply:
One corner case I have not considered, I think, is what would happen if all of the screenshots are just removed for a device type and language (or some of them are removed). I am not sure this implementation would just delete the missing ones, which would be the expected behaviour. I need to check. 😄

Reply:
Hmm... aren't we changing the meaning of the flag anyway? I mean, with the original implementation `:overwrite_screenshots` would effectively nuke all screenshots before an upload, while with this one there's no longer the option to do so. 

What you say is true, but I assumed that the original implementation of fastlane of the non-overwrite was simply flawed, and not an intentional use case - I can't see why someone would call fastlane to only upload new screenshots without deleting the old ones, especially considering that there's a strict cap on how many you can upload 🤔

Reply:
I think it depends on the use case. For people who are using the Fastlane CLI on a small scale, adding a single screenshot without touching anything else could be a genuine use case. With `:overwrite_screenshots` set to `true`, the person would need to supply all of the screenshots to the CLI. Not sure how it would work, maybe it is easier than it sounds, but I think supplying a single screenshot would be easier. 😃 

Regarding the fact whether we are changing it already or not, you are right. However, you'd need to assume someone would call the operation with no screenshots supplied in order to nuke everything. Otherwise, you always end up with the old/new screenshots in the end since they will be uploaded.

Reply:
If I manage to fix the case I mentioned previously, the original `:overwrite_screenshots` set to `true` effect would be available, if you just provide no screenshots. 🤷‍♂️

Reply:
However, we surely don't have any use for `:overwrite_screenshots` being set to `false` in our workflows. So, if we want to customise it more to our how we use it, the option you say is the better one for sure.

Reply:
> For people who are using the Fastlane CLI on a small scale, adding a single screenshot without touching anything else could be a genuine use case.

That is a possibility, although I'm not sure how common that would be. People generally use fastlane to automate their workflow, otherwise they'd just use App Store Connect. You put your screenshots in a folder, possibly generated automatically from tests, and [deliver](https://docs.fastlane.tools/actions/deliver/) takes care of uploading them for you. If that's the case, why would you want to upload an individual screenshot through fastlane? It'd not be faster than doing so from the portal _and_ it would require you to access the portal at some point anyway to delete existing screenshots (otherwise you run out of slots). 🤔 

> Otherwise, you always end up with the old/new screenshots in the end since they will be uploaded.
Also true, but again, in a normal scenario you'd have fastlane integrated with your continuous integration pipeline (like CircleCI), which means that you usually want to replace the old screenshots with the new ones (and not just to nuke them for the sake of it 😁). 

Ultimately, if we believe that both use cases are legitimate, then we're effectively adding a _third_ use case and we should treat it as such. For example, we could change the type of the `:overwrite_screenshots` option from boolean to enum and introduce three values: `:no`, `:all`, `:only_updated`, and treat its absence as the same thing as `:no`.

Reply:
So, in the end, I decided to disregard the `:overwrite_screenshots` flag all together for two reasons:
1) My implementation, that we delete screenshots only sometimes, is not something we need since we would always want to overwrite something and replace existing screenshots.
2) With the modification of the deletion algorithm (so that it deletes screenshots that need to be replaced but also ones that are straight up removed), the `nuke` option, where everything would be deleted and everything would be uploaded, would always end up with the same result in the end with many more operations involved.